### PR TITLE
I/6122: Handling spanned table cells in pasting scenarios

### LINKED ISSUE #6122: Handling spanned cell in pasting scenarios
## 📝 Provide a description of the new feature

This case requires #6120 & #6121 to proceed. Add tests and fix potential issues of pasting tables with spanned cells.

A table feature already provides a table layout post-fixer that handles edge cases in which such a broken layout could be created in the editor (most often during collaboration). However, the paste must be aware of such a condition and minimize the effect of a potentially broken layout by trimming the cells that extend over the rectangle selection.

### Cases to handle

*   [ ] selection has spans at the edges (might not work well or is blocked)
*   [ ] non-rectangular selection (might be tricky) - require to split table cell similarly when changing headings algorithm
*   [ ] non-rectangular pasted table

---

If you'd like to see this feature implemented, add a 👍 reaction to this post.

### LINKED ISSUE #6658: Setting column as header should split it col-spaned cells
## 📝 Provide detailed reproduction steps (if any)

1.  Merge two or more table cells at the beginning of a row.
2.  Set column heading on merged cell.

### ✔️ Expected result

The behavior should follow the logic for setting heading rows - it should split col-spanned cells which spans extends over the column heading section.

### ❌ Actual result

The restriction is not applied.

Note that creating an identical table is not possible using merge cells and merge (right/left) cell commands.

### Possible solution

For better UX when selection is in cells that are already merged I can see two options:

1.  We use it's `column` and split the merged cell.
2.  We set heading over columns that this cell spans.

<table><tbody><tr><td><strong>Case 1</strong>: split selected cell</td><td><strong>Case 2</strong>: Extend heading rows</td></tr><tr><td><figure class="image"></figure></td><td><figure class="image"></figure></td></tr></tbody></table>

For case 2: Any other merged cell (apart from selected) should not extend the heading section and should be split.

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by jodator**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Internal (table): Handle spanned table cells in pasting scenarios. Closes #6122.

Fix (table): Setting column as header will now split it col-spaned cells. Closes #6658.

---

### Additional information

This PR:

*   Introduces handling cases of spanned cells (both in selected table and pasted table) that have spans that exceeds rectangular selection.
*   Introduces fix to #6658 as the utils used were also created (they mirrors the logic of setting table header rows).
*   Handles a rectangular selection with cells spanned at the last row or column (like here: https://github.com/ckeditor/ckeditor5/pull/6786#issuecomment-627214164)
*   We also assume that pasted table is either valid or near valid. One case of invalid table structure was moved to a followup.

### CODE REVIEW

**Comment by jodator** [comment]
@niegowski it is ready for first review. Please keep in mind that some code was moved from other location (set header row command).

@ckeditor/qa-team - would be lovely to have another quick batch of pasting scenarios:

*   fixed issues from previous PR ([#6786](https://github.com/ckeditor/ckeditor5/pull/6786))
*   handled cases from the above mentioned PR:
    *   non rectangular selection handling (it should paste a table now)
    *   non-rectangular table in the clipboard (ie copied from an external editor)
    *   pasting table when selection is inside single cell (collapsed/not-collapsed, various block selected inside table cell) - the expected behaviour is to paste whole table from clipboard to a table and expand existing table to match dimensions.
*   still not covered (pasting smaller table than a selection)

**Comment by jodator** [comment]
@niegowski I'll handle the coverage drop also.

**Comment by FilipTokarski** [comment]
Not 100% sure if this is valid for this PR ( or rather I didn't catch it in the previous one ), but I'll share this here:  
- Case 1
19_table1

- Case 2
19_table2

In my view, the expected behaviour should be something like this:  
  
table_cases

File: packages/ckeditor5-table/src/tableclipboard.js [hunk 1]
```diff
@@ -349,3 +378,143 @@ function createLocationMap( table, width, height ) {
...
+// In th example above:
+// - Cell "02" which have `rowspan = 4` must be trimmed at first and at after last row.
+// - Cell "03" which have `rowspan = 2` and `colspan = 2` must be trimmed at first column and after last row.
+// - Cells "00", "03" & "30" which cannot be cut by this algorithm as they are outside the trimmed area.
+// - Cell "13" cannot be cut as it is inside the trimmed area.
+function splitCellsToRectangularSelection( selectedTableCells, writer ) {
+	const table = findAncestor( 'table', selectedTableCells[ 0 ] );
+
+	const rowIndexes = getRowIndexes( selectedTableCells );
+	const columnIndexes = getColumnIndexes( selectedTableCells );
```
Comment (niegowski) [inline packages/ckeditor5-table/src/tableclipboard.js:423]: We already have those indexes in place where we call this function.

**Comment by niegowski** [comment]
In my last commit I addressed most issues. 

Still to discuss:

*   [Naming convention for `startRow`/`endRow` vs `firstRow`/`lastRow`](https://github.com/ckeditor/ckeditor5/pull/7226#discussion_r427514641)
*   [Real dimensions of pasted table](https://github.com/ckeditor/ckeditor5/pull/7226#discussion_r427517663)
*   We still need to extract some of the logic to functions.

Also I didn't reviewed tests.

File: packages/ckeditor5-table/src/tableclipboard.js [hunk 2]
```diff
@@ -470,51 +471,38 @@ function doVerticalSplit( table, splitColumn, limitRows, writer ) {
...
-	const everyCellHasSingleRowspan = lastRowMap.every( ( { rowspan } ) => rowspan === 1 );
-
-	if ( !everyCellHasSingleRowspan ) {
-		const rowspanAdjustment = lastRowMap.pop().rowspan - 1;
-		return rowIndexes.last + rowspanAdjustment;
-	}
-
-	// Default to the last row index.
-	return rowIndexes.last;
+	return rowIndexes.last + Math.max( 1, ...lastRowMap.map( ( { rowspan } ) => rowspan ) ) - 1;
```
Comment (jodator) [inline packages/ckeditor5-table/src/tableclipboard.js:496]: @niegowski - please don't write everything in one line :(

File: packages/ckeditor5-table/src/tableclipboard.js [hunk 3]
```diff
@@ -470,51 +471,38 @@ function doVerticalSplit( table, splitColumn, limitRows, writer ) {
...
-	const everyCellHasSingleColspan = lastColumnMap.every( ( { colspan } ) => colspan === 1 );
-
-	if ( !everyCellHasSingleColspan ) {
-		const colspanAdjustment = lastColumnMap.pop().colspan - 1;
-		return columnIndexes.last + colspanAdjustment;
-	}
-
-	// Default to the last column index.
-	return columnIndexes.last;
+	return columnIndexes.last + Math.max( 1, ...lastColumnMap.map( ( { colspan } ) => colspan ) ) - 1;
```
Comment (jodator) [inline packages/ckeditor5-table/src/tableclipboard.js:507]: Same here - really. This doesn't read good :(

File: packages/ckeditor5-table/src/tableclipboard.js [hunk 4]
```diff
@@ -123,56 +129,78 @@ export default class TableClipboard extends Plugin {
...
-			let { first: firstColumnOfSelection, last: lastColumnOfSelection } = getColumnIndexes( selectedTableCells );
-			let { first: firstRowOfSelection, last: lastRowOfSelection } = getRowIndexes( selectedTableCells );
+			const columnIndexes = getColumnIndexes( selectedTableCells );
+			const rowIndexes = getRowIndexes( selectedTableCells );
+
+			let { last: lastColumnOfSelection, first: firstColumnOfSelection } = columnIndexes;
+			let { first: firstRowOfSelection, last: lastRowOfSelection } = rowIndexes;
 
-			if ( selectedTableCells.length == 1 ) {
+			const pasteHeight = tableUtils.getRows( pastedTable );
```
[the remark this answers is not in the document]
Comment (jodator) [reply packages/ckeditor5-table/src/tableclipboard.js:139]: Yeah, forgot that there can be garbage in the clipboard.

File: packages/ckeditor5-table/src/commands/setheadercolumncommand.js [hunk 5]
```diff
@@ -69,12 +69,21 @@ export default class SetHeaderColumnCommand extends Command {
...
 
+		const { first, last } = getColumnIndexes( selectedCells );
 		const headingColumnsToSet = this.value ? first : last + 1;
 
 		model.change( writer => {
-			const table = findAncestor( 'table', selectedCells[ 0 ] );
+			if ( headingColumnsToSet ) {
+				// Changing heading columns requires to check if any of a heading cell is overlapping vertically the table head.
+				// Any table cell that has a colspan attribute > 1 will not exceed the table head so we need to fix it in columns before.
+				const overlappingCells = getVerticallyOverlappingCells( table, headingColumnsToSet );
```
[the remark this answers is not in the document]
Comment (jodator) [reply packages/ckeditor5-table/src/commands/setheadercolumncommand.js]: Yeah I might get that 1:1 to "split cells vertically/horizontally" so it might be that it is the other way around.

File: packages/ckeditor5-table/src/tableclipboard.js [hunk 6]
```diff
@@ -123,56 +129,78 @@ export default class TableClipboard extends Plugin {
...
 
 				expandTableSize( selectedTable, lastRowOfSelection + 1, lastColumnOfSelection + 1, writer, tableUtils );
 			}
 
-			// Currently not handled. The selected table content should be trimmed to a rectangular selection.
-			// See: https://github.com/ckeditor/ckeditor5/issues/6122.
-			else if ( !isSelectionRectangular( selectedTableCells, tableUtils ) ) {
-				// @if CK_DEBUG // console.log( 'NOT IMPLEMENTED YET: Selection is not rectangular (non-mergeable).' );
+			let lastRowOfSelectionArea = lastRowOfSelection;
+			let lastColumnOfSelectionArea = lastColumnOfSelection;
```
[the remark this answers is not in the document]
Comment (jodator) [reply packages/ckeditor5-table/src/tableclipboard.js]: That was a though merge :/ I didn't review that fully afterwards, so you're right here.

**Comment by FilipTokarski** [comment]
If there are vertically and horizontally merged cells and they intersect each other on paste:
- first pasting case removes color from orange cell and splits it into 6 cells
- second case creates new column and moves 6th cell there
  
20_table1

**Comment by niegowski** [comment]
> If there are vertically and horizontally merged cells and they intersect each other on paste:
> 
> *   first pasting case removes color from orange cell and splits it into 6 cells
> *   second case creates new column and moves 6th cell there

Did you checked it after latest fixes?

**Comment by FilipTokarski** [comment]
Sorry, my bad 🤦 Splitting rows works fine now. Only the color issue still exists in some cases:
  
20_table2

**Comment by jodator** [comment]
> Did you checked it after latest fixes?

Please - if you fix something - add tests. I'm debugging what's happening as there was no new tests and I'm thinking why this is working now.

Test are needed for changed logic and bug fixes. I'll add those test.

File: packages/ckeditor5-table/src/tableclipboard.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (jodator) [reply packages/ckeditor5-table/src/tableclipboard.js:139]: I'd go with a follow up for that. It is also broken in the clipboard and might be a corner case (however will probably happen during testing phase :( )

[the remark this answers is not in the document]
Comment (jodator) [reply packages/ckeditor5-table/src/tableclipboard.js:139]: This might be tricky and a corner case: #7245.

File: packages/ckeditor5-table/src/tableclipboard.js [hunk 7]
```diff
@@ -123,56 +129,78 @@ export default class TableClipboard extends Plugin {
...
+			//
+			// Note: The table dimensions are established by the width of the first row and the total number of rows.
+			// It is possible to programmatically create a table that has rows which would have cells anchored beyond first row width but
+			// such table will not be created by other editing solutions.
+			const cropDimensions = {
+				startRow: 0,
+				startColumn: 0,
+				endRow: Math.min( selectionHeight - 1, pasteHeight - 1 ),
+				endColumn: Math.min( selectionWidth - 1, pasteWidth - 1 )
+			};
```
[the remark this answers is not in the document]
Comment (jodator) [reply packages/ckeditor5-table/src/tableclipboard.js:206]: That's a good point. Unfortunately we have both in some API points and I'm not sure which one is better.

**Comment by FilipTokarski** [comment]
### Steps to reproduce:
1. Create a 2x2 table and some another bigger table
2. Select all cells in 2x2 table and merge them into one
3. Now there are two scenarios: 
    - A ) select small table by holding shift and clicking the cell, copy and paste into bigger table
    - B ) select small table as a whole, copy and paste into bigger table

In scenario A pasted content is a single cell, whereas in B it remembers that initial table width was 2 columns. Not sure if this is correct, but it's a bit counterintuitive, because what you see when copying, is only a one-cell-table:
  
20_table3

File: packages/ckeditor5-table/src/utils.js [hunk 8]
```diff
@@ -264,11 +264,11 @@ export function isSelectionRectangular( selectedTableCells, tableUtils ) {
  *    4  │ o │ p │   │   │ q │
  *       └───┴───┴───┴───┴───┘
  *
- * will return cells: "j", "f", "k".
+ * will return slot info for cells: "j", "f", "k".
  *
  * @param {module:engine/model/element~Element} table The table to check.
  * @param {Number} overlapRow The index of the row to check.
- * @param {Number} [startRow=0] A row to start analysis if it is known where.
+ * @param {Number} [startRow=0] A row to start analysis. Use it when it is known that cells below that row will not overlap.
```
Comment (niegowski) [inline packages/ckeditor5-table/src/utils.js:271]:
```suggestion
 * @param {Number} [startRow=0] A row to start analysis. Use it when it is known that the cells above that row will not overlap.
```

**Comment by jodator** [comment]
> Sorry, my bad facepalm Splitting rows works fine now. Only the color issue still exists in some cases:

Moved here: #7246.

**Comment by jodator** [comment]
> Select all cells in 2x2 table and merge them into one

The bug is here :) Already reported: #6609. Would be nice to fix it eventually.
